scene/3d/chain_ik_3d.cpp

```diff
@@ -0,0 +1,484 @@
+	_make_simulation_dirty(p_index);
+	Skeleton3D *sk = get_skeleton();
+	if (sk && !chain_settings[p_index]->joints.is_empty()) {
+		_validate_axis(sk, p_index, chain_settings[p_index]->joints.size() - 1);
+	}
+	notify_property_list_changed();
+#ifdef TOOLS_ENABLED
+	update_gizmos();
+#endif // TOOLS_ENABLED
+}
+
+bool ChainIK3D::is_end_bone_extended(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, (int)settings.size(), false);
+	return chain_settings[p_index]->extend_end_bone;
+}
+
+void ChainIK3D::set_end_bone_direction(int p_index, BoneDirection p_bone_direction) {
+	ERR_FAIL_INDEX(p_index, (int)settings.size());
+	chain_settings[p_index]->end_bone_direction = p_bone_direction;
+	_make_simulation_dirty(p_index);
+	Skeleton3D *sk = get_skeleton();
+	if (sk && !chain_settings[p_index]->joints.is_empty()) {
+		_validate_axis(sk, p_index, chain_settings[p_index]->joints.size() - 1);
+	}
+#ifdef TOOLS_ENABLED
+	update_gizmos();
+#endif // TOOLS_ENABLED
+}
+
+SkeletonModifier3D::BoneDirection ChainIK3D::get_end_bone_direction(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, (int)settings.size(), BONE_DIRECTION_FROM_PARENT);
+	return chain_settings[p_index]->end_bone_direction;
+}
+
+void ChainIK3D::set_end_bone_length(int p_index, float p_length) {
+	ERR_FAIL_INDEX(p_index, (int)settings.size());
+	chain_settings[p_index]->end_bone_length = p_length;
+	_make_simulation_dirty(p_index);
+#ifdef TOOLS_ENABLED
+	update_gizmos();
+#endif // TOOLS_ENABLED
+}
+
+float ChainIK3D::get_end_bone_length(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, (int)settings.size(), 0);
+	return chain_settings[p_index]->end_bone_length;
+}
+
+// Individual joints.
+
+void ChainIK3D::set_joint_bone_name(int p_index, int p_joint, const String &p_bone_name) {
+	ERR_FAIL_INDEX(p_index, (int)settings.size());
+	LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	ERR_FAIL_INDEX(p_joint, (int)joints.size());
+	joints[p_joint].name = p_bone_name;
+	Skeleton3D *sk = get_skeleton();
+	if (sk) {
+		set_joint_bone(p_index, p_joint, sk->find_bone(joints[p_joint].name));
+	}
+}
+
+String ChainIK3D::get_joint_bone_name(int p_index, int p_joint) const {
+	ERR_FAIL_INDEX_V(p_index, (int)settings.size(), String());
+	const LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	ERR_FAIL_INDEX_V(p_joint, (int)joints.size(), String());
+	return joints[p_joint].name;
+}
+
+void ChainIK3D::set_joint_bone(int p_index, int p_joint, int p_bone) {
+	ERR_FAIL_INDEX(p_index, (int)settings.size());
+	LocalVector<BoneJoint> &joints = chain_settings[p_index]->joints;
+	ERR_FAIL_INDEX(p_joint, (int)joints.size());
+	joints[p_joint].bone = p_bone;
+	Skeleton3D *sk = get_skeleton();
+	if (sk) {
+		if (joints[p_joint].bone <= -1 || joints[p_joint].bone >= sk->get_bone_count()) {
+			WARN_PRINT("Joint bone index out of range!");
+			joints[p_joint].bone = -1;
+		} else {
+			joints[p_joint].name = sk->get_bone_name(joints[p_joint].bone);
+		}
+	}
```

Comment:
The setters behave differently depending on whether `get_skeleton()` is null.

I am worried there are cases which can desynchronize bone name/index

Reply:
Same as below https://github.com/godotengine/godot/pull/110120#discussion_r2462824011